### PR TITLE
Support for (de)?serialization of optionals

### PR DESCRIPTION
**Pull request by norrs**
Allow easy usage of optionals in APIs.

### CODE REVIEW

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 1]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
         RequestBody requestBody = JsonUtils.objectMapper.readValue(json, RequestBody.class);
         assertThat(requestBody.getField2(), is(nullValue()));
     }
+
+    @Test
+    public void jsonParserSerializedOptionalPresentAsT() throws JsonProcessingException {
```
Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:45]: jsonParserSerialize**s**OptionalPresentAsContainedValue?

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 2]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
...
+        String sut = JsonUtils.objectMapper.writeValueAsString(new TestObjectForOptionals("foo"));
```
Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:46]: Avoid magic strings. Use constants such as SAMPLE_VALUE

Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:46]: sut->actual

Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:46]: Do not use magic strings.

File: build.gradle [hunk 3]
```diff
@@ -56,7 +56,7 @@ dependencies {
     api group: 'org.zalando', name: 'jackson-datatype-problem', version: '0.24.0'
     runtimeOnly group: 'com.amazonaws', name: 'aws-java-sdk-api-gateway', version: awsSdkVersion
     implementation group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: jacksonVersion
-    runtimeOnly group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jdk8', version: jacksonVersion
+    implementation group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jdk8', version: jacksonVersion
```
Comment (axthosarouris) [inline build.gradle:59]: :+1:

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 4]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
         RequestBody requestBody = JsonUtils.objectMapper.readValue(json, RequestBody.class);
         assertThat(requestBody.getField2(), is(nullValue()));
     }
+
+    @Test
```
Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:44]: There are many magic string and a lot of replication in the Test Classes. I have summarized all my comments in the following suggestion. You can find more detail comments below.

```suggestion
   public static final JsonNode EMPTY_VALUE = null;
    public static final String SAMPLE_VALUE = "foo";
    public static final JsonNode JSON_OBJECT_WITH_VALUE = sampleJsonObjectWithSomeValue();
    public static final JsonNode JSON_OBJECT_WITHOUT_VALUE = sampleJsonObjectWithoutValue();


 @Test
    public void jsonParserSerializesOptionalPresentAsContainedObject() {
        JsonNode actual = serialize(objectWithSomeValue());
        assertThat(actual, is(equalTo(JSON_OBJECT_WITH_VALUE)));
    }

    @Test
    public void jsonParserSerializesOptionalNotPresentAsNull() {
        JsonNode actual = serialize(objectWithoutValue());
        assertThat(actual, is(equalTo(JSON_OBJECT_WITHOUT_VALUE)));
    }

    @Test
    public void jsonParserDeserializesOptionalNotPresent() {
        TestObjectForOptionals actual = deserialize(JSON_OBJECT_WITHOUT_VALUE);
        assertThat(actual.getTest(), is(equalTo(Optional.empty())));
        assertThat(actual.getTest(), isEmpty());
    }

    @Test
    public void jsonParserDeserializesOptionalPresent() {
        TestObjectForOptionals sut = deserialize(JSON_OBJECT_WITH_VALUE);
        assertThat(sut.getTest(), isPresentAnd(containsString(SAMPLE_VALUE)));
    }

    private TestObjectForOptionals objectWithoutValue() {
        return new TestObjectForOptionals(null);
    }

    private TestObjectForOptionals deserialize(JsonNode jsonObjectWithoutValue) {
        return JsonUtils.objectMapper.convertValue(jsonObjectWithoutValue,
            TestObjectForOptionals.class);
    }

 

    private TestObjectForOptionals objectWithSomeValue() {
        return new TestObjectForOptionals(SAMPLE_VALUE);
    }

    private <T> JsonNode serialize(T objectWithNullValue) {
        return JsonUtils.objectMapper.convertValue(objectWithNullValue, JsonNode.class);
    }

    private static JsonNode sampleJsonObjectWithSomeValue() {
        return JsonUtils.objectMapper.createObjectNode().put(JSON_KEY, SAMPLE_VALUE);
    }

    private static JsonNode sampleJsonObjectWithoutValue() {
        return JsonUtils.objectMapper.createObjectNode().set(JSON_KEY, EMPTY_VALUE);
    }

    private static class TestObjectForOptionals {

        private final String test;

        @JsonCreator
        public TestObjectForOptionals(@JsonProperty(JSON_KEY) String test) {
            this.test = test;
        }

        public Optional<String> getTest() {
            return Optional.ofNullable(test);
        }
    }
```

File: src/test/java/nva/commons/utils/JsonUtilsTest.java (hunk 2 above)
Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:46]: I think the system under testing is the library or the Json parser. Better if you called it actual?

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 5]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
...
+        public TestObjectForOptionals(@JsonProperty("test") String test) {
+            this.test = test;
+        }
+
+        public Optional<String> getTest() {
+            return Optional.ofNullable(test);
+        }
+    }
+
+    private static class TestObjectOptionalOfNullable {
```
Comment (axthosarouris) [inline src/test/java/nva/commons/utils/JsonUtilsTest.java:90]: The classes TestObjectForOptionalNullable and TestObjectOptionalOfEmpty are not necessary.

**Review by axthosarouris — CHANGES_REQUESTED** [review]
I lack a bit of context in the PR. Where are we going to serialize Optionals? Can you give an example?

File: src/test/java/nva/commons/utils/JsonUtilsTest.java (hunk 1 above)
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java (hunk 2 above)
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 6]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
         RequestBody requestBody = JsonUtils.objectMapper.readValue(json, RequestBody.class);
         assertThat(requestBody.getField2(), is(nullValue()));
     }
+
+    @Test
+    public void jsonParserSerializedOptionalPresentAsT() throws JsonProcessingException {
+        String sut = JsonUtils.objectMapper.writeValueAsString(new TestObjectForOptionals("foo"));
+        assertThat(sut, is(equalTo("{\n  \"test\" : \"foo\"\n}")));
```
[the remark this answers is not in the document]
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 7]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
...
+    }
+
+    @Test
+    public void jsonParserSerializedOptionalNotPresentAsNull() throws JsonProcessingException {
+        String sut = JsonUtils.objectMapper.writeValueAsString(new TestObjectForOptionals(null));
```
[the remark this answers is not in the document]
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java [hunk 8]
```diff
@@ -31,4 +40,66 @@ private void readFieldFromJson(String fileName) throws JsonProcessingException {
...
+        public Optional<String> getTest() {
+            return Optional.ofNullable(test);
+        }
+    }
+
+    private static class TestObjectOptionalOfNullable {
+        private String test;
+
+        public Optional<String> getTest() {
+            return Optional.ofNullable(test);
```
[the remark this answers is not in the document]
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java (hunk 2 above)
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

File: src/test/java/nva/commons/utils/JsonUtilsTest.java (hunk 5 above)
Comment (norrs) [reply src/test/java/nva/commons/utils/JsonUtilsTest.java]: @axthosarouris : Accepted your nice refactoring, should solve this.

**Review by axthosarouris — DISMISSED** [review]

File: src/main/java/nva/commons/utils/JsonUtils.java [hunk 9]
```diff
@@ -30,14 +31,16 @@ private static ObjectMapper createJsonParser() {
...
+            .registerModule(new Jdk8Module())
             .registerModule(emptyStringAsNullModule())
             .enable(DeserializationFeature.ACCEPT_EMPTY_STRING_AS_NULL_OBJECT)
             .enable(SerializationFeature.INDENT_OUTPUT)
             .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
             // We want date-time format, not unix timestamps
             .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
             // Ignore null fields
-            .setSerializationInclusion(Include.NON_NULL);
+            .setSerializationInclusion(Include.NON_NULL)
```
Comment (axthosarouris) [inline src/main/java/nva/commons/utils/JsonUtils.java:42]: do you need both? I think Include.NON_ABSENT contains Include.NON_NULL

**Review by axthosarouris — APPROVED** [review]

File: src/main/java/nva/commons/utils/JsonUtils.java (hunk 9 above)
Comment (norrs) [reply src/main/java/nva/commons/utils/JsonUtils.java]: @axthosarouris : You are correct, tests are passing when removing it, so I have pushed updates again. :+1:

**Review by axthosarouris — APPROVED** [review]

File: src/main/java/nva/commons/utils/JsonUtils.java (hunk 9 above)
Comment (axthosarouris) [reply src/main/java/nva/commons/utils/JsonUtils.java]: It is written in the javadoc :)

**Pull request merged**